pkg/replication/replicator.go

```diff
@@ -245,13 +251,47 @@ func (txr *TxReplicator) disconnect() {
 }
 
 func (txr *TxReplicator) fetchTX() ([]byte, error) {
-	exportTxStream, err := txr.client.ExportTx(txr.clientContext, &schema.ExportTxRequest{Tx: txr.nextTx})
+	precommitState, err := txr.db.CurrentPreCommitState()
+	if err != nil {
+		return nil, err
+	}
+
+	state := &schema.FollowerState{
+		UUID:              txr.uuid.String(), // TODO: analize if there is any advantage in sending as header
+		PrecommittedTxID:  precommitState.TxId,
+		PrecommittedTxAlh: precommitState.TxHash,
+	}
+
+	exportTxStream, err := txr.client.ExportTx(txr.clientContext, &schema.ExportTxRequest{
+		Tx:                  txr.nextTx,
+		FollowerState:       state,
+		IncludePreCommitted: true,
+	})
 	if err != nil {
 		return nil, err
 	}
 
 	receiver := txr.streamSrvFactory.NewMsgReceiver(exportTxStream)
-	return receiver.ReadFully()
+	bs, err := receiver.ReadFully()
+	if err != nil {
+		return nil, err
+	}
+
+	md := exportTxStream.Trailer()
+
+	if len(md.Get("may-commit-up-to-txid-bin")) > 0 && len(md.Get("may-commit-up-to-alh-bin")) > 0 {
+		mayCommitUpToTxID := binary.BigEndian.Uint64([]byte(md.Get("may-commit-up-to-txid-bin")[0]))
+
+		var mayCommitUpToAlh [sha256.Size]byte
+		copy(mayCommitUpToAlh[:], []byte(md.Get("may-commit-up-to-alh-bin")[0]))
+
+		err = txr.db.AllowCommitUpto(mayCommitUpToTxID, mayCommitUpToAlh)
```

Comment:
This currently fails, when master is started and replica connected, it is not able to fetch the transaction (tx not found error). The pre-committed state is not consistent with master. I assume this is because we are reading from the immustore before the tx has been added to it